source/windows/windows_pki_utils.c

```diff
@@ -181,31 +198,269 @@ int aws_import_trusted_certificates(
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_PKI,
+            "static: error creating a new rsa crypto context for key %s with errno %d",
+            uuid_str,
+            (int)GetLastError());
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto done;
+    }
+
+    if (!CryptImportKey(crypto_prov, key, decoded_len, 0, 0, &h_key)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_PKI,
+            "static: failed to import rsa key %s into crypto provider, error code %d",
+            uuid_str,
+            GetLastError());
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto done;
+    }
+
+    CRYPT_KEY_PROV_INFO key_prov_info;
+    AWS_ZERO_STRUCT(key_prov_info);
+    key_prov_info.pwszContainerName = uuid_wstr;
+    key_prov_info.dwProvType = PROV_RSA_FULL;
+    key_prov_info.dwKeySpec = AT_KEYEXCHANGE;
+
+    if (!CertSetCertificateContextProperty(*certs, CERT_KEY_PROV_INFO_PROP_ID, 0, &key_prov_info)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_PKI,
+            "static: error creating a new certificate context for key %s with errno %d",
+            uuid_str,
+            (int)GetLastError());
+        aws_raise_error(AWS_ERROR_SYS_CALL_FAILURE);
+        goto done;
+    }
+
+    result = AWS_OP_SUCCESS;
+
+done:
+
+    if (h_key != 0) {
+        CryptDestroyKey(h_key);
+    }
+
+    if (crypto_prov != 0) {
+        CryptReleaseContext(crypto_prov, 0);
+    }
+
+    return result;
+}
+
+#define ECC_256_MAGIC_NUMBER 0x20
+#define ECC_384_MAGIC_NUMBER 0x30
+
+static ULONG s_compute_ecc_key_type_from_private_key_size(size_t private_key_len) {
+    switch (private_key_len) {
+        case ECC_256_MAGIC_NUMBER:
+            return BCRYPT_ECDSA_PRIVATE_P256_MAGIC;
+        case ECC_384_MAGIC_NUMBER:
+            return BCRYPT_ECDSA_PRIVATE_P384_MAGIC;
+        default:
+            return BCRYPT_ECDSA_PRIVATE_P521_MAGIC;
+    }
+}
+
+static int s_cert_context_import_ecc_private_key(
+    PCCERT_CONTEXT *certs,
+    struct aws_allocator *allocator,
+    BYTE *key,
+    DWORD decoded_len,
+    char uuid_str[AWS_UUID_STR_LEN]) {
+
+    (void)decoded_len;
+
+    AWS_FATAL_ASSERT(certs != NULL);
+    const CERT_CONTEXT *cert_context = *certs;
+    AWS_FATAL_ASSERT(cert_context != NULL);
+
+    HCRYPTPROV crypto_prov = 0;
+    HCRYPTKEY h_key = 0;
+    BCRYPT_ECCKEY_BLOB *key_blob = NULL;
+    int result = AWS_OP_ERR;
+    SECURITY_STATUS status;
+
+    CRYPT_BIT_BLOB *public_key_blob = &cert_context->pCertInfo->SubjectPublicKeyInfo.PublicKey;
+    DWORD public_key_blob_length = public_key_blob->cbData;
+    if (public_key_blob_length == 0) {
+        AWS_LOGF_ERROR(AWS_LS_IO_PKI, "static: invalid zero-length ecc key data");
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        goto done;
+    }
+
+    public_key_blob_length--;
+    struct aws_byte_cursor public_blob_cursor = {
+        .ptr = public_key_blob->pbData + 1,
```

Comment:
Why is this +1? Could you add a comment?

Reply:
Now with a big ol' comment.  Not completely resolved, but I'm not going to try and manually edit the DER form of a certificate to test a hypothesis that is mostly of curiosity.